pallet/liquidity-pools/src/lib.rs

```diff
@@ -0,0 +1,1065 @@
+		type Call = Call<T>;
+
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			match call {
+				Call::rollover_unsigned { id, current_block } => {
+					let current_block_number = <frame_system::Pallet<T>>::block_number();
+					if &current_block_number < current_block {
+						return InvalidTransaction::Future.into()
+					}
+					ValidTransaction::with_tag_prefix("LiquidityPoolsChainWorker")
+						.priority(UNSIGNED_PRIORITY)
+						.and_provides(id)
+						.longevity(64_u64)
+						.propagate(true)
+						.build()
+				},
+				_ => InvalidTransaction::Call.into(),
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Calculates the reward amount for a user based on the amount staked in an asset,
+		/// the interest rate, and the already accumulated reward debt.
+		///
+		/// This function computes the reward in the asset's token based on the user's staked amount
+		/// and the specified interest rate. It then deducts any previously accumulated reward debt
+		/// from this amount. Finally, it converts the calculated reward to the equivalent amount in
+		/// native tokens, taking into account the difference in decimal places between the asset
+		/// token and the native token.
+		///
+		/// Parameters:
+		/// - `user_joined_amount`: The amount of the asset token that the user has staked in the
+		///   pool.
+		/// - `reward_debt`: The amount of reward that has already been accounted for the user,
+		///   which should be subtracted from the newly calculated reward.
+		/// - `interest_rate`: The interest rate for the pool, expressed in basis points.
+		/// - `interest_rate_base_point`: The divisor used to convert the basis point interest rate
+		///   into a proportion.
+		/// - `asset_decimals`: The number of decimal places used for the asset token.
+		/// - `native_decimals`: The number of decimal places used for the native token.
+		///
+		/// Returns:
+		/// - The calculated reward amount in native tokens, after adjusting for decimal places and
+		///   subtracting the reward debt.
+		pub fn calculate_reward(
+			user_joined_amount: Balance,
+			reward_debt: Balance,
+			interest_rate: u32,
+			interest_rate_base_point: u32,
+			asset_decimals: u8,
+			native_decimals: u8,
+		) -> Balance {
+			// Calculate reward in asset token
+			let mut reward = user_joined_amount
+				.saturating_mul(interest_rate.into())
+				.saturating_div(interest_rate_base_point.into());
+			// Remaining rewards
+			reward = reward.saturating_sub(reward_debt);
+
+			// Convert reward to native token based on decimals
+			if asset_decimals > native_decimals {
+				reward = reward
+					.saturating_div(10_u128.pow((asset_decimals - native_decimals).into()).into());
+			} else if asset_decimals < native_decimals {
+				reward = reward
+					.saturating_mul(10_u128.pow((native_decimals - asset_decimals).into()).into());
+			}
+
+			reward
+		}
+
+		pub fn get_vault_account(pool_id: T::PoolId) -> Option<T::AccountId> {
```

Comment:
In most cases where this function is called, we simply unwrap the result.
Implying this function should pretty much be infallible.
Can we make this infallible (internally) - and then change the function signature?

```suggestion
		pub fn get_vault_account(pool_id: T::PoolId) -> T::AccountId {
```